# Objective
To explore and visualize the dataset, build a linear regression model to predict the prices of used cars, and generate a set of insights and recommendations that will help the business.

# Key Requirement
come up with a pricing model that can effectively predict the price of used cars and can help the business in devising profitable strategies using differential pricing

# Data Dictionary

The data contains the different attributes of used cars sold in different locations. The detailed data dictionary is given below.

1. S.No.: Serial number
2. Name: Name of the car which includes brand name and model name
3. Location: Location in which the car is being sold or is available for purchase (cities)
4. Year: Manufacturing year of the car
5. Kilometers_driven: The total kilometers (a unit used to measure length or distance) driven in the car by the previous owner(s)
6. Fuel_Type: The type of fuel used by the car (Petrol, Diesel, Electric, CNG, LPG)
7. Transmission: The type of transmission used by the car (Automatic/Manual)
8. Owner: Type of ownership
9. Mileage: The standard mileage offered by the car company in kmpl or km/kg
10. Engine: The displacement volume of the engine in CC
11. Power: The maximum power of the engine in bhp
12. Seats: The number of seats in the car
13. New_Price: The price of a new car of the same model in INR Lakhs (1 Lakh INR = 100,000 INR)
14. Price: The price of the used car in INR Lakhs


# # Libraries to help with reading and manipulating data


In [1]:
import numpy as np
import pandas as pd
# Libraries to help with data visualization
import matplotlib.pyplot as plt
import seaborn as sns

# Libraries to help with data visualization
%matplotlib inline

# to split the data into train and test
from sklearn.model_selection import train_test_split

# to build linear regression_model
from sklearn.linear_model import LinearRegression

# to check model performance
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

# Removes the limit from the number of displayed columns and rows.
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 100)


# Reading Data from csv


In [2]:
#To connect to Google drive and use Google Colab 

from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
CarDataFromFile=pd.read_csv('/content/drive/MyDrive/used_cars_data.csv')

## Let creating working copy
carsData=CarDataFromFile.copy()

#  Data Exploration

In [4]:
# to check shape of data 
print(f'In Cars Data set there are {carsData.shape[0]} rows and {carsData.shape[1]} columns')

In Cars Data set there are 7253 rows and 14 columns


In [5]:
np.random.seed(1)
carsData.sample(10)

S.No.                                       Name    Location  Year  \
2397   2397             Ford EcoSport 1.5 Petrol Trend     Kolkata  2016   
3777   3777                     Maruti Wagon R VXI 1.2       Kochi  2015   
4425   4425                     Ford Endeavour 4x2 XLT   Hyderabad  2007   
3661   3661  Mercedes-Benz E-Class E250 CDI Avantgrade  Coimbatore  2016   
4514   4514       Hyundai Xcent 1.2 Kappa AT SX Option       Kochi  2016   
599     599             Toyota Innova Crysta 2.8 ZX AT  Coimbatore  2019   
186     186  Mercedes-Benz E-Class E250 CDI Avantgrade   Bangalore  2014   
305     305     Audi A6 2011-2015 2.0 TDI Premium Plus       Kochi  2014   
4582   4582                      Hyundai i20 1.2 Magna     Kolkata  2011   
5434   5434           Honda WR-V Edge Edition i-VTEC S       Kochi  2019   

      Kilometers_Driven Fuel_Type Transmission Owner_Type     Mileage  \
2397              21460    Petrol       Manual      First   17.0 kmpl   
3777              49818    Petrol       Manual      First   21.5 kmpl   
4425             130000    Diesel       Manual      First   13.1 kmpl   
3661              39753    Diesel    Automatic      First   13.0 kmpl   
4514              45560    Petrol    Automatic      First   16.9 kmpl   
599               40674    Diesel    Automatic      First  11.36 kmpl   
186               37382    Diesel    Automatic      First   13.0 kmpl   
305               61726    Diesel    Automatic      First  17.68 kmpl   
4582              36000    Petrol       Manual      First   18.5 kmpl   
5434              13913    Petrol       Manual      First   17.5 kmpl   

       Engine       Power  Seats  New_Price  Price  
2397  1497 CC  121.36 bhp    5.0       9.47   6.00  
3777  1197 CC   81.80 bhp    5.0       5.44   4.11  
4425  2499 CC     141 bhp    7.0      35.29   6.00  
3661  2143 CC   201.1 bhp    5.0      86.97  35.28  
4514  1197 CC      82 bhp    5.0       8.23   6.34  
599   2755 CC   171.5 bhp    7.0      28.05  24.82  
186   2143 CC   201.1 bhp    5.0      86.97  32.00  
305   1968 CC  174.33 bhp    5.0      67.10  20.77  
4582  1197 CC      80 bhp    5.0      10.25   2.50  
5434  1199 CC    88.7 bhp    5.0       9.36   8.20

In [6]:
carsData.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7253 entries, 0 to 7252
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   S.No.              7253 non-null   int64  
 1   Name               7253 non-null   object 
 2   Location           7253 non-null   object 
 3   Year               7253 non-null   int64  
 4   Kilometers_Driven  7253 non-null   int64  
 5   Fuel_Type          7253 non-null   object 
 6   Transmission       7253 non-null   object 
 7   Owner_Type         7253 non-null   object 
 8   Mileage            7251 non-null   object 
 9   Engine             7207 non-null   object 
 10  Power              7078 non-null   object 
 11  Seats              7200 non-null   float64
 12  New_Price          7253 non-null   float64
 13  Price              6019 non-null   float64
dtypes: float64(3), int64(3), object(8)
memory usage: 793.4+ KB


In [7]:
carsData.describe(include='all').T

count unique                     top  freq          mean  \
S.No.              7253.0    NaN                     NaN   NaN        3626.0   
Name                 7253   2041  Mahindra XUV500 W8 2WD    55           NaN   
Location             7253     11                  Mumbai   949           NaN   
Year               7253.0    NaN                     NaN   NaN   2013.365366   
Kilometers_Driven  7253.0    NaN                     NaN   NaN  58699.063146   
Fuel_Type            7253      5                  Diesel  3852           NaN   
Transmission         7253      2                  Manual  5204           NaN   
Owner_Type           7253      4                   First  5952           NaN   
Mileage              7251    450               17.0 kmpl   207           NaN   
Engine               7207    150                 1197 CC   732           NaN   
Power                7078    385                  74 bhp   280           NaN   
Seats              7200.0    NaN                     NaN   NaN      5.279722   
New_Price          7253.0    NaN                     NaN   NaN     21.307322   
Price              6019.0    NaN                     NaN   NaN      9.479468   

                            std     min      25%      50%      75%        max  
S.No.               2093.905084     0.0   1813.0   3626.0   5439.0     7252.0  
Name                        NaN     NaN      NaN      NaN      NaN        NaN  
Location                    NaN     NaN      NaN      NaN      NaN        NaN  
Year                   3.254421  1996.0   2011.0   2014.0   2016.0     2019.0  
Kilometers_Driven  84427.720583   171.0  34000.0  53416.0  73000.0  6500000.0  
Fuel_Type                   NaN     NaN      NaN      NaN      NaN        NaN  
Transmission                NaN     NaN      NaN      NaN      NaN        NaN  
Owner_Type                  NaN     NaN      NaN      NaN      NaN        NaN  
Mileage                     NaN     NaN      NaN      NaN      NaN        NaN  
Engine                      NaN     NaN      NaN      NaN      NaN        NaN  
Power                       NaN     NaN      NaN      NaN      NaN        NaN  
Seats                   0.81166     0.0      5.0      5.0      5.0       10.0  
New_Price             24.256314    3.91     7.88     11.3    21.69      375.0  
Price                 11.187917    0.44      3.5     5.64     9.95      160.0

 Approach  🇰
 Let us first do so univariable analysis 

 Split Name into multiple columns so we can see if we can find  company name and derive some conclusion on it 
 We can drop Price  missing values 
 Convert data type for Mileage , Engine and Power 
 Convert  Year using to year 
  convery Fuel type , trnasmission , owen_type to categorical

In [8]:
carsData.duplicated().sum()

0

S.No. is not required  as it is just a counter .

In [9]:
carsData.drop('S.No.',axis=1,inplace=True)

In [10]:
carsData.reset_index(inplace=True,drop=True)

In [11]:
carsData.isnull().sum().sort_values(ascending=False)

Price                1234
Power                 175
Seats                  53
Engine                 46
Mileage                 2
Name                    0
Location                0
Year                    0
Kilometers_Driven       0
Fuel_Type               0
Transmission            0
Owner_Type              0
New_Price               0
dtype: int64

In [12]:
carsData.dropna(subset=['Price'], inplace=True)

In [13]:
carsData.isnull().sum()

Name                   0
Location               0
Year                   0
Kilometers_Driven      0
Fuel_Type              0
Transmission           0
Owner_Type             0
Mileage                2
Engine                36
Power                143
Seats                 42
New_Price              0
Price                  0
dtype: int64

Mileage , Engine and Power arshould be  numeric but they are appearing as Objects .Lets explore it

In [14]:
carsData[['Mileage','Power','Engine']].tail(2000)

Mileage       Power   Engine
4019   21.9 kmpl   88.76 bhp  1396 CC
4020   25.2 kmpl      74 bhp  1248 CC
4021   11.7 kmpl     178 bhp  1796 CC
4022  14.28 kmpl  138.03 bhp  1798 CC
4023  22.32 kmpl  126.32 bhp  1582 CC
...          ...         ...      ...
6014   28.4 kmpl      74 bhp  1248 CC
6015   24.4 kmpl      71 bhp  1120 CC
6016   14.0 kmpl     112 bhp  2498 CC
6017   18.9 kmpl    67.1 bhp   998 CC
6018  25.44 kmpl    57.6 bhp   936 CC

[2000 rows x 3 columns]

In [15]:
position_cols=['Mileage','Power','Engine']

In [16]:
 ## Function to remove string  like km/kg', 'kmpl' from  columns Mileage , Engine and Power and convert it to Float

In [17]:
def covertToNumeric(colname):
  if isinstance(colname, str):
   return float((colname.split(' ')[0]))
  else:
    return np.nan

    

In [18]:
for colname in position_cols :
  carsData[colname]=carsData[colname].apply(covertToNumeric)

In [19]:
carsData.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6019 entries, 0 to 6018
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Name               6019 non-null   object 
 1   Location           6019 non-null   object 
 2   Year               6019 non-null   int64  
 3   Kilometers_Driven  6019 non-null   int64  
 4   Fuel_Type          6019 non-null   object 
 5   Transmission       6019 non-null   object 
 6   Owner_Type         6019 non-null   object 
 7   Mileage            6017 non-null   float64
 8   Engine             5983 non-null   float64
 9   Power              5876 non-null   float64
 10  Seats              5977 non-null   float64
 11  New_Price          6019 non-null   float64
 12  Price              6019 non-null   float64
dtypes: float64(6), int64(2), object(5)
memory usage: 658.3+ KB


In [20]:
carsData.describe(include='all').T

count unique                     top  freq          mean  \
Name                 6019   1876  Mahindra XUV500 W8 2WD    49           NaN   
Location             6019     11                  Mumbai   790           NaN   
Year               6019.0    NaN                     NaN   NaN   2013.358199   
Kilometers_Driven  6019.0    NaN                     NaN   NaN  58738.380296   
Fuel_Type            6019      5                  Diesel  3205           NaN   
Transmission         6019      2                  Manual  4299           NaN   
Owner_Type           6019      4                   First  4929           NaN   
Mileage            6017.0    NaN                     NaN   NaN     18.134961   
Engine             5983.0    NaN                     NaN   NaN    1621.27645   
Power              5876.0    NaN                     NaN   NaN     113.25305   
Seats              5977.0    NaN                     NaN   NaN      5.278735   
New_Price          6019.0    NaN                     NaN   NaN     21.499855   
Price              6019.0    NaN                     NaN   NaN      9.479468   

                            std     min      25%      50%      75%        max  
Name                        NaN     NaN      NaN      NaN      NaN        NaN  
Location                    NaN     NaN      NaN      NaN      NaN        NaN  
Year                   3.269742  1998.0   2011.0   2014.0   2016.0     2019.0  
Kilometers_Driven  91268.843206   171.0  34000.0  53000.0  73000.0  6500000.0  
Fuel_Type                   NaN     NaN      NaN      NaN      NaN        NaN  
Transmission                NaN     NaN      NaN      NaN      NaN        NaN  
Owner_Type                  NaN     NaN      NaN      NaN      NaN        NaN  
Mileage                4.582289     0.0    15.17    18.15     21.1      33.54  
Engine               601.355233    72.0   1198.0   1493.0   1984.0     5998.0  
Power                 53.874957    34.2     75.0     97.7    138.1      560.0  
Seats                   0.80884     0.0      5.0      5.0      5.0       10.0  
New_Price             24.332922    3.91     7.88     11.3    21.77      375.0  
Price                 11.187917    0.44      3.5     5.64     9.95      160.0

# To analyze Car Name
1. First we will check Name which conains both Company Name and Model in single 
fields as how many cars are sold for specific Name.
2. In next step we will split Car Name into Company Name and Model name and check value counts for company name and model  individually .
3. Based on  Step 1 and Step 2 output we will decide whether we shoudl use Name as single field or should we first separte and covert it to categorical variable for further analysis . 

In [78]:
CarNameDf=carsData['Name'].value_counts().sort_values(ascending =False)

In [93]:
CarNameDf.head(100)

Mahindra XUV500 W8 2WD                        49
Maruti Swift VDI                              45
Honda City 1.5 S MT                           34
Maruti Swift Dzire VDI                        34
Maruti Swift VDI BSIV                         31
Maruti Ritz VDi                               30
Hyundai i10 Sportz                            30
Toyota Fortuner 3.0 Diesel                    29
Honda Amaze S i-Dtech                         27
Hyundai Grand i10 Sportz                      27
Maruti Ertiga VDI                             25
Honda Brio S MT                               25
Honda City 1.5 V MT                           24
Maruti WagonR VXI                             23
Hyundai Verna 1.6 SX                          23
Maruti WagonR LXI                             22
Maruti Alto LXi                               21
Hyundai EON Era Plus                          20
Maruti Ertiga ZDI                             20
Skoda Superb Elegance 1.8 TSI AT              20
BMW 3 Series 320d   

There are total 1876 Names (combinations of Company Name and Model Name occuring in Data set ) with maximum Cars being sold by Mahindra of model XUV500 W8 2WD  followed by Maruti Swift VDI .

Car Name contains Model and Company Name so let  us split these two and see how each  Company and model is occuring in dataset.

In [121]:
CarCompanyNameModel=carsData['Name'].str.split(" ",n=2,expand=True)

In [123]:
CarCompanyNameModel.columns=['CompanyName','Model','Ignore']

we will see how many cars are sold by each company based on it's name and store it in CompanyNameCounts data Frame.

In [124]:
CompanyNameCounts=CarCompanyNameModel['CompanyName'].value_counts()
CompanyNameCounts.head(50)

Maruti           1211
Hyundai          1107
Honda             608
Toyota            411
Mercedes-Benz     318
Volkswagen        315
Ford              300
Mahindra          272
BMW               267
Audi              236
Tata              186
Skoda             173
Renault           145
Chevrolet         121
Nissan             91
Land               60
Jaguar             40
Fiat               28
Mitsubishi         27
Mini               26
Volvo              21
Porsche            18
Jeep               15
Datsun             13
Force               3
ISUZU               2
Smart               1
Ambassador          1
Isuzu               1
Bentley             1
Lamborghini         1
Name: CompanyName, dtype: int64

We can see that Maximum occurance is of company Name Maruti and Hundai which depicts that maximum cars are being sold are for Maruti and Hundai  

we will see how many cars are sold by each company based on it's naModel and store it in ModelCounts data Frame.

In [125]:
ModelCounts=CarCompanyNameModel['Model'].value_counts().sort_values(ascending=False)

In [127]:
ModelCounts.head(50)

Swift        353
City         270
i20          247
Verna        170
Innova       164
i10          156
Grand        156
WagonR       154
Polo         151
Alto         143
XUV500       113
Vento        107
Amaze        106
New          102
Figo         101
Fortuner     101
3             98
Creta         93
E-Class       88
Duster        83
A4            78
Santro        78
Ertiga        75
Ciaz          71
Corolla       68
5             68
Ritz          64
Baleno        63
Etios         61
EON           61
Scorpio       61
Rover         60
Brio          60
Xcent         59
Celerio       58
Jazz          58
Ecosport      53
Superb        49
A6            49
Rapid         47
Beat          46
Vitara        43
EcoSport      42
Fiesta        40
KWID          40
Indica        40
Endeavour     39
Q7            38
SX4           37
Micra         33
Name: Model, dtype: int64

When we checked top 100 we could see  maximum times there is entry for model XUV500 W8 2WD , followed by Swift VDI.

## Treatment for original Name Column from CarsData dataset
1.  We have analyzed Car Name as single string and also by separing it between Company name and Model . 
2. When we analyzed Name as single field there are 1876 unique combinations
3. Hence For this analysis we will first add Company Name and Model as new columns in carsData data Frame . 
4. For both Company Name and Model  we will  First create two  Lookup  list  for for Company Name and one  for Model which will have top 20 copany name and Model which we want to keep in data set and rest of the  company names and Model name we will code it as "Other" 


In [128]:
carsData['CompanyName']=CarCompanyNameModel['CompanyName']
carsData['Model']=CarCompanyNameModel['Model']


In [141]:
## Lookup  Master for Company Name  and Model which will top top 20 Company Names and Model Name respectively 
Lookup_Master_companyName = CompanyNameCounts[CompanyNameCounts.values >= 20].index.tolist()
Lookup_Master_companyModel = ModelCounts[ModelCounts.values >= 35].index.tolist()


In [146]:
car_temp = []

for i in range(carsData.shape[0]): 
   
    check = 0  # to see of variable in data frame is matching with Lookup field
    for item in Lookup_Master_companyName:  
        if item in carsData['CompanyName'].values[i]  :
            car_temp.append(item)
            check = 1
    if check == 0:  
        car_temp.append("Others")
carsData["revised_CompanyName"] = [item for item in car_temp]


In [147]:
carsData["revised_CompanyName"].value_counts()

Maruti           1211
Hyundai          1107
Honda             608
Toyota            411
Mercedes-Benz     318
Volkswagen        315
Ford              300
Mahindra          272
BMW               267
Audi              236
Tata              186
Skoda             173
Renault           145
Chevrolet         121
Nissan             91
Land               60
Others             56
Jaguar             40
Fiat               28
Mitsubishi         27
Mini               26
Volvo              21
Name: revised_CompanyName, dtype: int64

In [150]:
car_temp = []

for i in range(carsData.shape[0]): 
   
    check = 0  # to see of variable in data frame is matching with Lookup field
    for item in Lookup_Master_companyModel:  
        if item in carsData['Model'].values[i]  and check == 0:
            car_temp.append(item)
            check = 1
    if check == 0:  
        car_temp.append("Others")
carsData["revised_CompanyModel"] = [item for item in car_temp]

In [151]:
carsData.describe(include='all').T

count unique                     top  freq  \
Name                    6019   1876  Mahindra XUV500 W8 2WD    49   
Location                6019     11                  Mumbai   790   
Year                  6019.0    NaN                     NaN   NaN   
Kilometers_Driven     6019.0    NaN                     NaN   NaN   
Fuel_Type               6019      5                  Diesel  3205   
Transmission            6019      2                  Manual  4299   
Owner_Type              6019      4                   First  4929   
Mileage               6017.0    NaN                     NaN   NaN   
Engine                5983.0    NaN                     NaN   NaN   
Power                 5876.0    NaN                     NaN   NaN   
Seats                 5977.0    NaN                     NaN   NaN   
New_Price             6019.0    NaN                     NaN   NaN   
Price                 6019.0    NaN                     NaN   NaN   
CompanyName             6019     31                  Maruti  1211   
Model                   6019    212                   Swift   353   
Names_Model             6019     50                'Others'  4595   
revised_CompanyName     6019     22                  Maruti  1211   
revised_CompanyModel    6019     50                  Others  1346   

                              mean           std     min      25%      50%  \
Name                           NaN           NaN     NaN      NaN      NaN   
Location                       NaN           NaN     NaN      NaN      NaN   
Year                   2013.358199      3.269742  1998.0   2011.0   2014.0   
Kilometers_Driven     58738.380296  91268.843206   171.0  34000.0  53000.0   
Fuel_Type                      NaN           NaN     NaN      NaN      NaN   
Transmission                   NaN           NaN     NaN      NaN      NaN   
Owner_Type                     NaN           NaN     NaN      NaN      NaN   
Mileage                  18.134961      4.582289     0.0    15.17    18.15   
Engine                  1621.27645    601.355233    72.0   1198.0   1493.0   
Power                    113.25305     53.874957    34.2     75.0     97.7   
Seats                     5.278735       0.80884     0.0      5.0      5.0   
New_Price                21.499855     24.332922    3.91     7.88     11.3   
Price                     9.479468     11.187917    0.44      3.5     5.64   
CompanyName                    NaN           NaN     NaN      NaN      NaN   
Model                          NaN           NaN     NaN      NaN      NaN   
Names_Model                    NaN           NaN     NaN      NaN      NaN   
revised_CompanyName            NaN           NaN     NaN      NaN      NaN   
revised_CompanyModel           NaN           NaN     NaN      NaN      NaN   

                          75%        max  
Name                      NaN        NaN  
Location                  NaN        NaN  
Year                   2016.0     2019.0  
Kilometers_Driven     73000.0  6500000.0  
Fuel_Type                 NaN        NaN  
Transmission              NaN        NaN  
Owner_Type                NaN        NaN  
Mileage                  21.1      33.54  
Engine                 1984.0     5998.0  
Power                   138.1      560.0  
Seats                     5.0       10.0  
New_Price               21.77      375.0  
Price                    9.95      160.0  
CompanyName               NaN        NaN  
Model                     NaN        NaN  
Names_Model               NaN        NaN  
revised_CompanyName       NaN        NaN  
revised_CompanyModel      NaN        NaN

In [152]:
carsData["revised_CompanyModel"].value_counts()

Others       1346
Swift         353
City          270
i20           247
Verna         170
Innova        164
Grand         161
i10           156
Polo          155
WagonR        154
3             144
Alto          143
5             124
XUV500        113
Vento         107
Amaze         106
New           102
Fortuner      101
Figo          101
Creta          93
E-Class        88
Duster         83
Santro         78
A4             78
Ertiga         75
Ciaz           71
Corolla        68
Ritz           64
Baleno         63
Etios          61
Scorpio        61
EON            61
Brio           60
Rover          60
Xcent          59
Celerio        58
Jazz           58
Ecosport       53
Superb         49
A6             49
Rapid          47
Beat           46
Vitara         43
EcoSport       42
Fiesta         40
Indica         40
KWID           40
Endeavour      39
Q7             38
SX4            37
Name: revised_CompanyModel, dtype: int64

In [ ]:
carsData['Transmission']=carsData['Transmission'].astype('category')

In [ ]:
carsData['Fuel_Type']=carsData['Fuel_Type'].astype('category')

In [ ]:
carsData['Owner_Type']=carsData['Owner_Type'].astype('category')

In [ ]:
carsData.info()

In [ ]:
carsData['Seats'].fillna(carsData['Seats'].mean(),inplace=True)


In [ ]:
carsData.describe(include='all').T

In [ ]:
carsData['Mileage'].fillna(carsData['Mileage'].mean(),inplace=True)

In [ ]:
carsData['Engine'].fillna(carsData['Engine'].median(),inplace=True)

In [ ]:
carsData['Power'].fillna(carsData['Power'].median(),inplace=True)

In [ ]:
carsData.isnull().sum()

In [ ]:
carsData.describe(include='all').T

In [49]:
kk=carsData.Name.value_counts()

In [105]:
cc=carsData['Name'].str.split(" ",n=2,expand=True)

In [114]:
ss=cc[1].value_counts().sort_values(ascending=False)

In [120]:
ss[ss.values >= 100]

Swift       353
City        270
i20         247
Verna       170
Innova      164
i10         156
Grand       156
WagonR      154
Polo        151
Alto        143
XUV500      113
Vento       107
Amaze       106
New         102
Figo        101
Fortuner    101
Name: 1, dtype: int64